vermouth/processors/do_links.py

```diff
@@ -261,36 +262,147 @@ def _build_link_interaction_from(molecule, interaction, match):
     )
     return new_interaction
 
-
 class DoLinks(Processor):
     """
     Apply Links, taken from a molecule's force field, to the molecule.
     """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.applied_links = defaultdict(dict)
+        self.current_match = None
+        self.current_link = None
+        self.nodes_to_remove = []
+
+    def replace_attributes(self, molecule, match):
+        """
+        Given any nodes in `self.current_link` have the 'replace'
+        attribute, replace the attribute in molecule with the new
+        value. If the atomname is set to None stage the node to be
+        removed.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        for node, node_attrs in self.current_link.nodes.items():
+            if 'replace' in node_attrs:
+                if node_attrs['replace'].get('atomname', False) is None:
+                    self.nodes_to_remove.append(match[node])
+                else:
+                    node_mol = molecule.nodes[match[node]]
+                    node_mol.update(node_attrs['replace'])
+        return molecule
+
+    def add_interactions(self, molecule, match):
+        """
+        Add interactions from link to the `self.applied_interactions` dict.
+        We temporarily store them in this dict since links can remove and or
+        replace interactions but the built in replace_interaction function in
+        molecule requires looping over the complete interactions dict, which
+        is slow. Having interactions in a dict with key also takes care of all
+        symmetry issues.
```

Comment:
Well yes I suggested that about 2 years ago and it was denied. Now we cannot change it otherwise I've to change all polyply and other program codes that iterate over interaction dicts. I veto that for now even though in general I agree. Perhaps we can put it as long term improvement.

Reply:
Fair point (: 

If we decide to change this (at some point) we should provide an interface that behaves the way it does now, at least temporarily.

Reply:
agreed! @pckroon would you be open to doing this in this PR? I had a look at the polyply code and if we provide a temp interface then when I've time later in the year I can adopt the code quite easily

Reply:
Also on this point I think we need a custom reverse comparison method